src/main/kotlin/no/nav/personbruker/dittnav/eventaggregator/done/DoneEventService.kt

```diff
@@ -2,82 +2,71 @@ package no.nav.personbruker.dittnav.eventaggregator.done
-        val eventMetricsProbe: EventMetricsProbe
+        private val doneRepository: DoneRepository,
+        private val eventMetricsProbe: EventMetricsProbe
 ) : EventBatchProcessorService<Done> {
 
     private val log: Logger = LoggerFactory.getLogger(DoneEventService::class.java)
 
     override suspend fun processEvents(events: ConsumerRecords<Nokkel, Done>) {
+        val successfullyTransformedEvents = mutableListOf<no.nav.personbruker.dittnav.eventaggregator.done.Done>()
+        val problematicEvents = mutableListOf<ConsumerRecord<Nokkel, Done>>()
         events.forEach { event ->
             try {
                 eventMetricsProbe.reportEventSeen(DONE, event.systembruker)
-                processDoneEvent(event)
+                val internalEvent = DoneTransformer.toInternal(event.getNonNullKey(), event.value())
+                successfullyTransformedEvents.add(internalEvent)
                 eventMetricsProbe.reportEventProcessed(DONE, event.systembruker)
+
             } catch (e: NokkelNullException) {
                 eventMetricsProbe.reportEventFailed(DONE, event.systembruker)
+                problematicEvents.add(event)
                 log.warn("Eventet manglet nøkkel. Topic: ${event.topic()}, Partition: ${event.partition()}, Offset: ${event.offset()}", e)
+
             } catch (e: Exception) {
                 eventMetricsProbe.reportEventFailed(DONE, event.systembruker)
+                problematicEvents.add(event)
                 log.warn("Transformasjon av done-event fra Kafka feilet.", e)
             }
         }
+        val groupedDoneEvents = groupDoneEventsByAssociatedEventType(successfullyTransformedEvents)
+        writeDoneEventsToCache(groupedDoneEvents)
+        kastExceptionHvisMislykkedeTransformasjoner(problematicEvents)
     }
 
-    private suspend fun processDoneEvent(event: ConsumerRecord<Nokkel, Done>) {
-        val doneEntity = DoneTransformer.toInternal(event.getNonNullKey(), event.value())
-        val brukernotifikasjoner = database.dbQuery { getAllBrukernotifikasjonFromView() }
-        val foundEvent: Brukernotifikasjon? = brukernotifikasjoner.find { isEventInCache(it, doneEntity) }
-        if (foundEvent != null) {
-            log.info("Fant matchende event for Done-event med eventId: ${foundEvent.eventId}, produsent: ${foundEvent.produsent}, type: ${foundEvent.type}")
-            flagEventAsInactive(foundEvent, doneEntity)
-        } else {
-            database.dbQuery { createDoneEvent(doneEntity) }
-            log.info("Fant ikke matchende event for done-event med eventId ${doneEntity.eventId}, produsent: ${doneEntity.produsent}, eventTidspunkt: ${doneEntity.eventTidspunkt}. Skrev done-event til cache")
-        }
+    private suspend fun groupDoneEventsByAssociatedEventType(successfullyTransformedEvents: MutableList<no.nav.personbruker.dittnav.eventaggregator.done.Done>): DoneBatchProcessor {
+        val brukernotifikasjoner = doneRepository.fetchBrukernotifikasjonerFromView()
```

Comment:
Her kan vi eventuelt optimalisere med å hente ut kun aktive eventer. Eller er det kanskje dumt, fordi det da kan bli liggende done-eventer i ventetabellen hvis et done event mottas to ganger. Det unngår vi med å alltid ta med alle eventer.

Reply:
Har vi noen ide om hvor stor gevinst vi får ved kun å hente ut aktive eventer?

Reply:
Jeg tipper at gevinsten er rimelig stor, fordi det over tid vil være mye færre aktive eventer enn det er eventer totalt. Eneste ulempen med å gjøre det slik er det jeg nevnte om done-eventer som kommer flere ganger.
Men vi kan jo kanskje løse dette ved at vi ved førstegangsprosessering av done-eventer sjekker etter aktive eventer, hvis vi ikke får treff så havner den i ventetabellen. Den periodiske jobben som behandler ventetabellen, den kan loope gjennom alle eventer.
Slik jeg ser det så får vi da det beste av to verdener 😄